File: ChatAI/Java/Web/SpringBoot/src/main/java/com/portfolio/chatai/provider/OpenAiCompatibleChatProvider.java
package com.portfolio.chatai.provider;

import com.portfolio.chatai.model.ChatChoice;
import com.portfolio.chatai.model.ChatRequest;
import com.portfolio.chatai.model.ChatResponse;
import com.portfolio.chatai.model.ChatUsage;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Component
public class OpenAiCompatibleChatProvider implements IChatProvider {

    private final RestTemplate restTemplate;
    private final ObjectMapper mapper;
    private final String baseUrl;
    private final String apiKey;

    @Autowired
    public OpenAiCompatibleChatProvider(RestTemplate restTemplate, ObjectMapper mapper) {
        this.restTemplate = restTemplate;
        this.mapper = mapper;
        this.baseUrl = System.getenv().getOrDefault("OPENAI_BASE_URL", "https://api.openai.com/v1");
        this.apiKey = System.getenv().getOrDefault("OPENAI_API_KEY", "");
    }

    @Override
    public ChatResponse completeChat(ChatRequest request) {
        String model = request.getModel() != null ? request.getModel() : System.getenv().getOrDefault("CHAT_MODEL", "gpt-4o-mini");
        Double temperature = request.getTemperature() != null ? request.getTemperature() : Double.parseDouble(System.getenv().getOrDefault("CHAT_TEMPERATURE", "0.7"));
        Integer maxTokens = request.getMax_tokens() != null ? request.getMax_tokens() : Integer.parseInt(System.getenv().getOrDefault("CHAT_MAX_TOKENS", "1024"));

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("model", model);
        List<Map<String, String>> messages = request.getMessages().stream()
                .map(m -> Map.of("role", m.getRole(), "content", m.getContent()))
                .toList();
        payload.put("messages", messages);
        payload.put("temperature", temperature);
        payload.put("max_tokens", maxTokens);

        String url = baseUrl.endsWith("/") ? baseUrl + "chat/completions" : baseUrl + "/chat/completions";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        if (apiKey != null && !apiKey.isEmpty()) {
            headers.setBearerAuth(apiKey);
        }

        ResponseEntity<String> response;
        try {
            response = restTemplate.postForEntity(url, new HttpEntity<>(payload, headers), String.class);
        } catch (RestClientException e) {
            throw new RuntimeException("Provider error: " + e.getMessage());
        }

        return parseChatResponse(response.getBody(), model);
    }

    @SuppressWarnings("unchecked")
    private ChatResponse parseChatResponse(String body, String model) {
        ChatResponse resp = new ChatResponse();
        resp.setModel(model);
        if (body == null || body.isEmpty()) {
            resp.setId("");
            resp.setChoices(List.of(emptyChoice()));
            resp.setUsage(new ChatUsage());
            return resp;
        }
        try {
            Map<String, Object> data = mapper.readValue(body, Map.class);
            resp.setId((String) data.getOrDefault("id", ""));
            List<Map<String, Object>> choices = (List<Map<String, Object>>) data.getOrDefault("choices", List.of());
            List<ChatChoice> out = new ArrayList<>();
            for (Map<String, Object> c : choices) {
                Object msgObj = c.get("message");
                if (msgObj instanceof Map) {
                    Map<String, Object> msg = (Map<String, Object>) msgObj;
                    ChatChoice choice = new ChatChoice();
                    choice.setRole((String) msg.getOrDefault("role", "assistant"));
                    choice.setContent((String) msg.getOrDefault("content", ""));
                    out.add(choice);
                }
            }
            if (out.isEmpty()) {
                out.add(emptyChoice());
            }
            resp.setChoices(out);
            Map<String, Object> usageMap = (Map<String, Object>) data.getOrDefault("usage", Map.of());
            ChatUsage usage = new ChatUsage();
            usage.setPrompt_tokens(((Number) usageMap.getOrDefault("prompt_tokens", 0)).intValue());
            usage.setCompletion_tokens(((Number) usageMap.getOrDefault("completion_tokens", 0)).intValue());
            usage.setTotal_tokens(((Number) usageMap.getOrDefault("total_tokens", 0)).intValue());
            resp.setUsage(usage);
        } catch (Exception e) {
            resp.setId("");
            resp.setChoices(List.of(emptyChoice()));
            resp.setUsage(new ChatUsage());
        }
        return resp;
    }

    private static ChatChoice emptyChoice() {
        ChatChoice choice = new ChatChoice();
        choice.setRole("assistant");
        choice.setContent("");
        return choice;
    }
}

File: ChatAI/Elixir/Web/Phoenix/src/lib/chatai/providers/open_ai_compatible_chat_provider.ex
defmodule ChatAI.Providers.OpenAiCompatibleChatProvider do
  @moduledoc """
  OpenAI-compatible chat provider: `POST {base_url}/chat/completions` with
  `Authorization: Bearer {key}`.
  """
  @behaviour ChatAI.Providers.IChatProvider
  defstruct []

  alias ChatAI.{ChatRequest, ChatResponse}

  @impl true
  def complete_chat(%ChatRequest{} = request) do
    base_url = env("OPENAI_BASE_URL", "https://api.openai.com/v1")
    api_key = env("OPENAI_API_KEY", "")

    model = request.model || env("CHAT_MODEL", "gpt-4o-mini")
    temperature = request.temperature || parse_float(env("CHAT_TEMPERATURE", "0.7"))
    max_tokens = request.max_tokens || parse_int(env("CHAT_MAX_TOKENS", "1024"))
    timeout = ChatAI.Providers.Util.timeout_ms()

    url = base_url |> String.trim_trailing("/") |> Kernel.<>("/chat/completions")

    payload = %{
      model: model,
      messages: Enum.map(request.messages, &%{role: &1.role, content: &1.content}),
      temperature: temperature,
      max_tokens: max_tokens
    }

    headers = [
      {"content-type", "application/json"},
      {"authorization", "Bearer " <> api_key}
    ]

    case ChatAI.Providers.HttpClient.post_json(url, headers, payload, timeout) do
      {:ok, body} -> {:ok, parse_response(body, model)}
      {:error, msg} -> {:error, msg}
    end
  end

  defp parse_response(body, model) do
    case Jason.decode(body) do
      {:ok, data} ->
        choices =
          case data["choices"] do
            list when is_list(list) ->
              Enum.map(list, fn c ->
                msg = c["message"] || %{}
                %{role: msg["role"] || "assistant", content: msg["content"] || ""}
              end)

            _ ->
              [ChatResponse.empty_choice()]
          end

        usage = data["usage"] || %{}

        %ChatResponse{
          id: data["id"] || "",
          provider: "openai",
          model: model,
          choices: choices,
          usage: %{
            prompt_tokens: usage["prompt_tokens"] || 0,
            completion_tokens: usage["completion_tokens"] || 0,
            total_tokens: usage["total_tokens"] || 0
          }
        }

      {:error, _} ->
        %ChatResponse{
          id: "",
          provider: "openai",
          model: model,
          choices: [ChatResponse.empty_choice()],
          usage: %{}
        }
    end
  end

  defp env(key, default), do: System.get_env(key, default)

  defp parse_float(s) do
    case Float.parse(s) do
      {f, _} -> f
      :error -> 0.7
    end
  end

  defp parse_int(s) do
    case Integer.parse(s) do
      {i, _} -> i
      :error -> 1024
    end
  end
end
